### PR TITLE
ui: Change Password Feature

### LINKED ISSUE #79: Change password option
In GitLab by @jzakotnik on Feb 1, 2019, 10:21

This is important for user management or e.g. changing password at first login

### LINKED ISSUE #325: ui: Change password feature
Follow up to #79 - the UI part will be tracked in this issue. 

Todos: 

- [ ] UI component & E2E test - Profile screen? Where should user data be displayed? Who can update passwords?
- [ ] Documentation
- [ ] Changelog

### PR DESCRIPTION
**Pull request by mathiashoeld**
# Changes

A user can now change the password for each user, where he/she has the permission to do so (including his/her own password). The password can be changed on the user overview page. 
The user has to enter his/her own password first as security measure. 

Closes #325 

# How to test

There are new E2E tests which specifically test this feature. Otherwise, the password can be changed manually by clicking on the edit symbol next to the permissions symbol, entering the current password and then entering the new one.

### CODE REVIEW

File: api/src/user_list.ts [hunk 1]
```diff
@@ -62,6 +63,7 @@ interface ExposedIdentity {
   displayName: string;
   organization?: string;
   isGroup: boolean;
+  permissions: object;
```
Comment (gonzochic) [inline api/src/user_list.ts]: Use the permissions type

File: frontend/src/languages/portuguese.js [hunk 2]
```diff
@@ -81,7 +81,15 @@ const pt = {
     new_group: "Novo grupo de usuários",
     add_group: "Criar novo grupo",
     edit_group: "Editar grupo",
-    edit_permissions_for: "Editar permissões para"
+    edit_permissions_for: "Editar permissões para",
+    current_user_password: "Portuguese: Your password",
+    new_user_password: "Portuguese: New password",
```
Comment (gonzochic) [inline frontend/src/languages/portuguese.js]: Google Translate here

File: frontend/src/pages/Common/Password.js [hunk 3]
```diff
@@ -30,6 +30,7 @@ const Password = ({ password, storePassword, failed, nextBestAction, id, ...prop
     <div style={styles.container}>
       <PasswordIcon style={styles.icon} />
       <TextField
+        data-test="password-textfield"
```
Comment (gonzochic) [inline frontend/src/pages/Common/Password.js]: Could be problematic if you have mutliple password components on the screen
```suggestion
        data-test={props["data-test"] || "password-textfield"}
```

File: frontend/src/pages/Users/UserPasswordChangeDialogContent.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+          </div>
+        </CardContent>
+      </Card>
+      <Card style={styles.card}>
+        <CardHeader subheader={formatString(strings.users.type_new_password, userId)} />
+        <CardContent>
+          <div style={styles.container}>
+            <TextField
+              data-test="new-password-textfield"
+              style={{ width: "50%" }}
```
Comment (gonzochic) [inline frontend/src/pages/Users/UserPasswordChangeDialogContent.js]: use styles variable

File: frontend/src/pages/Users/Users.js [hunk 5]
```diff
@@ -41,7 +42,7 @@ const Users = props => {
   const onClick = () => (tabIndex === 0 ? showDashboardDialog("addUser") : showDashboardDialog("addGroup"));
   const permissionIconDisplayed = allowedIntents.includes("global.listPermissions");
   return (
-    <div id="userdashboard" style={styles.container}>
+    <div id="userdashboard" data-test="userdashboard" style={styles.container}>
```
Comment (gonzochic) [inline frontend/src/pages/Users/Users.js]: do we still need the id "userdashboard" here?

File: frontend/src/pages/Users/UsersTable.js [hunk 6]
```diff
@@ -1,15 +1,17 @@
...
 import TableHead from "@material-ui/core/TableHead";
 import TableRow from "@material-ui/core/TableRow";
-import TableCell from "@material-ui/core/TableCell";
-import TableBody from "@material-ui/core/TableBody";
-import Paper from "@material-ui/core/Paper";
-import _sortBy from "lodash/sortBy";
 import PermissionIcon from "@material-ui/icons/LockOpen";
-import IconButton from "@material-ui/core/IconButton";
+import _sortBy from "lodash/sortBy";
+import React from "react";
```
Comment (gonzochic) [inline frontend/src/pages/Users/UsersTable.js]: Check sorting

File: frontend/src/pages/Users/UsersTable.js [hunk 7]
```diff
@@ -20,7 +22,14 @@ const sortUsers = users => {
...
-const UsersTable = ({ classes, users, permissionIconDisplayed, showDashboardDialog, userId, allowedIntents }) => {
+const UsersTable = ({
+  classes,
+  users,
+  permissionIconDisplayed,
+  showDashboardDialog,
+  showChangeUserPasswordDialog,
+  userId
+}) => {
   const sortedUsers = sortUsers(users.filter(u => u.isGroup !== true)); //&& u.id !== userId));
```
Comment (gonzochic) [inline frontend/src/pages/Users/UsersTable.js]: What does the comment mean?

File: frontend/src/pages/Users/UserPasswordChangeDialogContent.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+          <div style={styles.container}>
+            <TextField
+              data-test="new-password-confirmation-textfield"
+              style={{ width: "50%" }}
+              label={strings.users.new_user_password_confirmation}
+              value={newPasswordConfirmation}
+              margin="normal"
+              error={!newPasswordsMatch}
+              id="newPasswordConfirmation"
+              onChange={event => storeNewPasswordConfirmation(event.target.value)}
```
Comment (gonzochic) [inline frontend/src/pages/Users/UserPasswordChangeDialogContent.js]: Is this necessary on change?

File: api/src/user_list.ts (hunk 1 above)
Comment (mathiashoeld) [reply api/src/user_list.ts]: done, changed to: 
``` 
  permissions: Permissions;
```

File: frontend/src/languages/portuguese.js (hunk 2 above)
Comment (mathiashoeld) [reply frontend/src/languages/portuguese.js]: done, also for french

File: frontend/src/pages/Users/UserPasswordChangeDialogContent.js (hunk 4 above)
Comment (mathiashoeld) [reply frontend/src/pages/Users/UserPasswordChangeDialogContent.js]: done

File: frontend/src/pages/Users/Users.js (hunk 5 above)
Comment (mathiashoeld) [reply frontend/src/pages/Users/Users.js]: no, removed

File: frontend/src/pages/Users/UsersTable.js (hunk 6 above)
Comment (mathiashoeld) [reply frontend/src/pages/Users/UsersTable.js]: used automatic sorting

File: frontend/src/pages/Users/UsersTable.js (hunk 7 above)
Comment (mathiashoeld) [reply frontend/src/pages/Users/UsersTable.js]: removed

File: frontend/src/pages/Users/UserPasswordChangeDialogContent.js (hunk 8 above)
Comment (mathiashoeld) [reply frontend/src/pages/Users/UserPasswordChangeDialogContent.js]: changed to the following implementation: 
- handle `onChange` event of `password`, `newPassword`, `newPasswordConfirmation` locally (with hooks)
- store the value in the redux store with the `onBlur` event
- enable submit as soon as one of the new password fields is not empty

File: frontend/src/pages/Common/Password.js (hunk 3 above)
Comment (mathiashoeld) [reply frontend/src/pages/Common/Password.js]: done, also refactored `<Password>` component to be able to re-use it in the dialog

**Pull request merged**